pkg/agent/openflow/pipeline_windows.go

```diff
@@ -0,0 +1,255 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openflow
+
+import (
+	"net"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/config"
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow/cookie"
+	binding "github.com/vmware-tanzu/antrea/pkg/ovs/openflow"
+)
+
+const (
+	markTrafficFromBridge = 5
+
+	// ctZoneSNAT is only used on Windows and only when AntreaProxy is enabled.
+	// When a Pod access a ClusterIP Service, and the IP of the selected endpoint
+	// is not in "cluster-cidr". The request packets need to be SNAT'd(set src IP to local Node IP)
+	// after have been DNAT'd(set dst IP to endpoint IP).
+	// For example, the endpoint Pod may run in hostNetwork mode and the IP of the endpoint
+	// will is the current Node IP.
+	// We need to use a different ct_zone to track the SNAT'd connection because OVS
+	// does not support doing both DNAT and SNAT in the same ct_zone.
+	//
+	// An example of the connection is a Pod accesses kubernetes API service:
+	// Pod --> DNAT(CtZone) --> SNAT(ctZoneSNAT) --> Endpoint(API server NodeIP)
+	// Pod <-- unDNAT(CtZone) <-- unSNAT(ctZoneSNAT) <-- Endpoint(API server NodeIP)
+	ctZoneSNAT = 0xffdc
+
+	// snatDefaultMark indicates the packet should be SNAT'd with the default
+	// SNAT IP (the Node IP).
+	snatDefaultMark = 0b1
+
+	// snatCTMark indicates SNAT is performed for packets of the connection.
+	snatCTMark = 0x40
+)
+
+var (
+	// snatMarkRange takes the 17th bit of register marksReg to indicate if
+	// the packet needs to be SNATed with Node's IP or not.
+	snatMarkRange = binding.Range{17, 17}
+)
+
+// uplinkSNATFlows installs flows for traffic from the uplink port that help
+// the SNAT implementation of the external traffic.
+func (c *client) uplinkSNATFlows(localSubnet net.IPNet, category cookie.Category) []binding.Flow {
+	ctStateNext := dnatTable
+	if c.enableProxy {
+		ctStateNext = endpointDNATTable
+	}
+	bridgeOFPort := uint32(config.BridgeOFPort)
+	flows := []binding.Flow{
+		// Mark the packet to indicate its destination MAC should be
+		// rewritten to the real MAC in the L3Forwarding table, if the
+		// packet is a reply to a local Pod from an external address.
+		c.pipeline[conntrackStateTable].BuildFlow(priorityHigh).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchCTMark(snatCTMark, nil).
+			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
+			// return packets to remote Pods will be handled by L3Forwarding flows.
+			MatchDstIPNet(localSubnet).
```

Comment:
I notice this condition is new, can a packet to remote Pods get to L3Forwarding flows? It looks like the following flowwith normal priority will output the packet to bridge port?

Reply:
Humm.. yes, such pakcets will be output to br-int, which we would not want to avoid asymmetric return path.

In this case, probably no harm to set macRewriteMark for all packets, though it is not needed for packets to remote Nodes.

Reply:
Updated the flow.